### PR TITLE
Patch paths; closes #123

### LINKED ISSUE #123: Update paths
Should we keep them immutable like routes?
- No

Consider a log for audit
- We should have audit

Change the format of the payload: 
```
{
    name: ...
    eskip: ... (optional - not needed for delete)
    timestamp: ...
    type: update | create | delete
}
```

Update the swagger spec.

### PR DESCRIPTION
**Pull request by NorthFury**

### CODE REVIEW

File: src/it/scala/org/zalando/spearheads/innkeeper/dao/PathsPostgresRepoSpec.scala [hunk 1]
```diff
@@ -1,13 +1,16 @@
 package org.zalando.spearheads.innkeeper.dao
 
 import java.time.LocalDateTime
+
```
Comment (danpersa) [inline src/it/scala/org/zalando/spearheads/innkeeper/dao/PathsPostgresRepoSpec.scala]: you shouldn't add more spaces...

File: src/it/scala/org/zalando/spearheads/innkeeper/routes/PatchPathsSpec.scala [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        val token = "token--employees-route.write_strict"
+
+        it("should return the 403 Forbidden status") {
+          val response = PathsSpecsHelper.patchSlashPaths(1L, pathPatchHostIdsJsonString, token)
+          response.status should be(StatusCodes.Forbidden)
+          entityString(response).parseJson.convertTo[Error].errorType should be("TNF")
+        }
+      }
+
+      describe("when a token without admin privileges is provided and owned_by_team is defined") {
```
Comment (danpersa) [inline src/it/scala/org/zalando/spearheads/innkeeper/routes/PatchPathsSpec.scala]: good one 👍

File: src/it/scala/org/zalando/spearheads/innkeeper/routes/PostPathsSpec.scala [hunk 3]
```diff
@@ -52,38 +52,38 @@ class PostPathsSpec extends FunSpec with BeforeAndAfter with Matchers {
           path.hostIds should be(hostIds)
         }
       }
-    }
 
-    describe("when a token with the admin scope is provided") {
```
Comment (danpersa) [inline src/it/scala/org/zalando/spearheads/innkeeper/routes/PostPathsSpec.scala]: why is this file reformatted?

Comment (NorthFury) [inline src/it/scala/org/zalando/spearheads/innkeeper/routes/PostPathsSpec.scala]: Please check individual commits for this pull request. There were two specs hanging around when they should have been in the `describe("success")` block.

File: src/main/scala/org/zalando/spearheads/innkeeper/dao/PathsRepo.scala [hunk 4]
```diff
@@ -11,4 +15,5 @@ trait PathsRepo {
   def selectAll: DatabasePublisher[PathRow]
   def selectByOwnerTeamAndUri(ownedByTeam: Option[String], uri: Option[String]): DatabasePublisher[PathRow]
   def pathWithUriExists(uri: String): Future[Boolean]
+  def patch(id: Long, pathPatch: PathPatch, updatedAt: LocalDateTime): Future[Option[PathRow]]
```
Comment (danpersa) [inline src/main/scala/org/zalando/spearheads/innkeeper/dao/PathsRepo.scala]: The names in the Repos should be in the database domain and not http domain. I would rename this to update.

File: src/main/scala/org/zalando/spearheads/innkeeper/api/JsonProtocols.scala [hunk 5]
```diff
@@ -126,19 +126,38 @@ object JsonProtocols {
...
     "uri",
     "host_ids",
     "owned_by_team",
     "created_by",
-    "created_at")
+    "created_at",
+    "updated_at"
+  )
+
+  implicit object RouteChangeTypeFormat extends RootJsonFormat[RouteChangeType] {
```
Comment (danpersa) [inline src/main/scala/org/zalando/spearheads/innkeeper/api/JsonProtocols.scala]: You should write some unit tests for this in the `JsonProtocolsSpec`

File: src/main/scala/org/zalando/spearheads/innkeeper/routes/PatchPaths.scala [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+          hasOneOfTheScopes(authenticatedUser, reqDesc, scopes.WRITE) {
+            logger.debug(s"patch /paths non-admin team $team")
+
+            if (pathPatch.ownedByTeam.isDefined) {
+              reject(PathOwnedByTeamAuthorizationRejection(reqDesc))
+            } else {
+              patchPathRoute(id, pathPatch, reqDesc)
+            }
+          } ~
+            hasAdminAuthorization(authenticatedUser, team, reqDesc, scopes)(teamService) {
```
Comment (danpersa) [inline src/main/scala/org/zalando/spearheads/innkeeper/routes/PatchPaths.scala]: I have the feeling that there should not be a new line here :)

File: src/main/scala/org/zalando/spearheads/innkeeper/services/EskipRouteService.scala [hunk 7]
```diff
@@ -6,7 +6,7 @@ import akka.NotUsed
 import akka.stream.scaladsl.Source
 import com.google.inject.Inject
 import org.zalando.spearheads.innkeeper.api.JsonProtocols._
-import org.zalando.spearheads.innkeeper.api.{EskipRoute, EskipRouteWrapper, NewRoute, RouteName, RouteOut}
```
Comment (danpersa) [inline src/main/scala/org/zalando/spearheads/innkeeper/services/EskipRouteService.scala]: You should use explicit imports whenever possible. You should instruct your IDE to use the explicit version.

File: src/main/scala/org/zalando/spearheads/innkeeper/services/EskipRouteService.scala [hunk 8]
```diff
@@ -72,11 +72,22 @@ class EskipRouteService @Inject() (routesRepo: RoutesRepo, routeToEskipTransform
 
     Source.fromPublisher(streamOfRows.mapResult {
       case (routeRow, pathRow) =>
+        val (routeChangeType, timestamp) = if (routeRow.deletedAt.isDefined) {
```
Comment (danpersa) [inline src/main/scala/org/zalando/spearheads/innkeeper/services/EskipRouteService.scala]: I would replace this with pattern matching. The code looks more like Java code than Scala code to me.

```scala
routeRow.deletedAt match {
  case Some(deletedAt) =>
    RouteChangeType.Delete -> deletedAt
  case _ =>
    if (pathRow.updatedAt ...)
```

File: src/main/scala/org/zalando/spearheads/innkeeper/services/PathsService.scala [hunk 9]
```diff
@@ -5,7 +5,7 @@ import java.time.LocalDateTime
 import akka.NotUsed
 import akka.stream.scaladsl.Source
 import com.google.inject.Inject
-import org.zalando.spearheads.innkeeper.api.{PathIn, PathOut, TeamName, UserName}
+import org.zalando.spearheads.innkeeper.api._
```
Comment (danpersa) [inline src/main/scala/org/zalando/spearheads/innkeeper/services/PathsService.scala]: Again, please don't remove implicit imports

**Comment by danpersa** [comment]
The ticket also #123 also mentions that you should update the swagger file

File: src/main/scala/org/zalando/spearheads/innkeeper/services/EskipRouteService.scala (hunk 8 above)
Comment (NorthFury) [inline src/main/scala/org/zalando/spearheads/innkeeper/services/EskipRouteService.scala]: :+1:

File: src/main/scala/org/zalando/spearheads/innkeeper/api/JsonProtocols.scala (hunk 5 above)
Comment (NorthFury) [inline src/main/scala/org/zalando/spearheads/innkeeper/api/JsonProtocols.scala]: done

File: src/main/scala/org/zalando/spearheads/innkeeper/services/EskipRouteService.scala (hunk 7 above)
Comment (NorthFury) [inline src/main/scala/org/zalando/spearheads/innkeeper/services/EskipRouteService.scala]: will be fixed in another pull request

**Comment by danpersa** [comment]
:+1:
